Add nested fragments

elidhu opened the pull request:
I have added nested fragment parsing. I have not had to touch any tests except:

- Removing the "double open" test as it is no longer an error.
- Adding a "halt while open" test that I noticed was missing.
- Adding a "nested fragment" test.

The original behaviour should be identical as far as the tests go.

The new behaviour can be seen in the example I added. The most notable part is the handling of nested fragment tags when outputting a fragment that has children. Nested tags are excluded as if they weren't there.

Let me know your thoughts.

Thanks!

nickpascucci commented on src/lib.rs:
```diff
@@ -155,118 +155,106 @@ pub fn extract_fragments(
 
+    if !fragment_stack.is_empty() {
+        return Err(FileError {
+            err_type: ParseError::DoubleOpen,
```
**request:** Can you rename this error variant to reflect the changed semantics? Perhaps something like `UnclosedFragment`.

elidhu replied:
Great pickup!

nickpascucci commented on src/lib.rs:
```diff
@@ -742,6 +730,60 @@ def main():
+
+        let fragments = fragments.expect("Expected no parse errors");
+        assert!(
+            fragments.len() == 3,
```
```suggestion
            fragments.len() == 3, // Two fragments in addition to the full file
```

nickpascucci commented on src/lib.rs:
```diff
@@ -742,6 +730,60 @@ def main():
+            "    Start of outer
+        Start of inner
+        End of inner
+    End of outer",
```
```suggestion
    End of outer",
            "Expected nested fragment markers to be removed"
```

nickpascucci reviewed (changes requested):
Thanks for your PR! This looks good, nice work.

Before I approve, can I ask you to update the README with:

- A mention that fragments can be nested
- Removing the "future work" item
- Adding yourself to the "Contributors" section

elidhu commented:
Addressed all of the above, and updated the README.

Thanks for taking the time to review!

elidhu commented:
I did have the thought while I was making the changes that this adds support for "nested" fragments, not necessarily "overlapping". This would required some sort of ID on the closing tag (I guess).

I'm not sure if that was an important distinction

nickpascucci commented:
Yes, good point — I don't think it's a hugely important distinction, and as you said, once the fragment structure is no longer tree-like it becomes necessary to annotate the closing tags. This seems like a good next step, adding a nice feature without complecting the system too much.

elidhu committed:
feat: add nested fragment extraction

Signed-off-by: Kevin Glasson <kevinglasson@gmail.com>

elidhu committed:
feat: add example for testing

Signed-off-by: Kevin Glasson <kevinglasson@gmail.com>

elidhu committed:
feat: add nested to check-examples

Signed-off-by: Kevin Glasson <kevinglasson@gmail.com>

elidhu committed:
fix: correct nested example

Signed-off-by: Kevin Glasson <kevinglasson@gmail.com>

elidhu committed:
fix: address PR comments

- Rename error variant DoubleOpen -> UnclosedFragment.
- Add clarifying comments to nested fragment test.

Signed-off-by: Kevin Glasson <kevinglasson@gmail.com>

elidhu committed:
feat: update README

Signed-off-by: Kevin Glasson <kevinglasson@gmail.com>

elidhu commented:
RIght, had to fix all my signatures. It's good-to-go now!

elidhu commented:
Would you be able to get this merged and cut a release? 🙏

nickpascucci commented:
@elidhu sorry for the wait — I've just published v0.3.0 which includes your change 🎉
